infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -18,54 +18,102 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
 import           Pos.Util.Util               (leftToPanic)
 
 import           Pos.Slotting.MemState.Class (MonadSlotsData (..))
-import           Pos.Slotting.Types          (EpochSlottingData (..), SlottingData (..))
+import           Pos.Slotting.Types          (EpochSlottingData (..))
 
 -- | Approximate current slot using outdated slotting data.
-approxSlotUsingOutdated :: MonadSlotsData m => EpochIndex -> Timestamp -> m SlotId
-approxSlotUsingOutdated penult t = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
-    let epochStart = esdStartDiff sdLast `addTimeDiffToTimestamp` systemStart
+approxSlotUsingOutdated
+    :: (MonadSlotsData m)
+    => Timestamp
+    -> m SlotId
+approxSlotUsingOutdated t = do
+
+    systemStart            <- getSystemStartM
+    (currentEpochIndex ,_) <- getCurrentNextEpochIndexM
+    (_, nextSlottingData)  <- getCurrentNextEpochSlottingDataM
```

Comment:
And I thought about this, this won't help. It will help with other things and clean some things up nicely, I have an idea how to fix it - instead of `IO` use `STM` and receive all the required data in one go.